### DISCUSSION TITLE
Add features via additional sensors to MJXDD001 / MJXDD01SYL (yeelink.light.ceiling22)

### DISCUSSION THREAD

**Discussion by B48D81EFCC** [syssi/esphome-yeelight-ceiling-light#81] [Ideas]
Hello,

I want to put some I2C sensors into the lamp housing. Sensors to measure temperature, humidity, air pressure and/or brightness.

Unfortunately, the built-in board has very limited GPIOs, but for I2C just two additional GPIOs are needed.
I am wondering if the the TX/RX pins could be used for I2C.

Alternatively, a pin-compatible replacement board could be used. Any suggestions?

**Comment by B48D81EFCC** [comment]
Okay, I looked a bit closer to the PCB and I found 3 GPIOs labeled with R G B.
I guess these pins are not used in this lamp. I guess I could use these as well.

**Reply by syssi** [reply]
Good catch! I guess you are on the right track. Just assign the `i2c` component to these GPIOs.

**Comment by syssi** [comment]
Most Yeelight devices come equipped with a EEPROM accessible via i2c. In other words: There is a i2c bus already. If you are able to identify the EEPROM you could attach the additional i2c devices here.

**Reply by syssi** [reply]
This is what I'm talking about:

https://github.com/syssi/esphome-yeelight-ceiling-light/blob/main/yeelight_light_lamp9.yaml#L123-L128

**Reply by B48D81EFCC** [reply]
Oh interesting. But does the lamp use this bus already?
Is I2C Multi-Master capable? Can the bus have multiple masters?

**Reply by syssi** [reply]
No. The stock firmware uses the i2c bus to store some encryption keys. The ESP32 is the master and most of the time a single device (the EEPROM) is available on the bus. Just the bslamp2 is a bit special here because it has some touch controls etc.

**Reply by B48D81EFCC** [reply]
Okay, got it! Thanks.

**Comment by B48D81EFCC** [comment]
Some findings so far using GPIO26, GPIO27 and GPIO33.

When I use GPIO33 (SDA) and GPIO26 (SCL) for the bus, apparently some devices are found.
```
i2c:
  sda: GPIO33
  scl: GPIO26
  scan: true
```

```
[10:39:05][C][i2c.idf:075]: I2C Bus:
[10:39:05][C][i2c.idf:076]:   SDA Pin: GPIO33
[10:39:05][C][i2c.idf:077]:   SCL Pin: GPIO26
[10:39:05][C][i2c.idf:078]:   Frequency: 50000 Hz
[10:39:05][C][i2c.idf:084]:   Recovery: bus successfully recovered
[10:39:05][I][i2c.idf:094]: Results from i2c bus scan:
[10:39:05][I][i2c.idf:100]: Found i2c device at address 0x0E
[10:39:05][I][i2c.idf:100]: Found i2c device at address 0x12
[10:39:05][I][i2c.idf:100]: Found i2c device at address 0x17
[10:39:05][I][i2c.idf:100]: Found i2c device at address 0x20
[10:39:05][I][i2c.idf:100]: Found i2c device at address 0x24
[10:39:05][I][i2c.idf:100]: Found i2c device at address 0x3C
[10:39:05][I][i2c.idf:100]: Found i2c device at address 0x3D
[10:39:05][I][i2c.idf:100]: Found i2c device at address 0x42
[10:39:05][I][i2c.idf:100]: Found i2c device at address 0x49
[10:39:05][I][i2c.idf:100]: Found i2c device at address 0x58
[10:39:05][I][i2c.idf:100]: Found i2c device at address 0x74
```

Any other combination of the 3 GPIOs did not find any devices:
```
[10:45:05][C][i2c.idf:090]:   Recovery: failed, SDA is held low on the bus
[10:45:05][I][i2c.idf:094]: Results from i2c bus scan:
[10:45:05][I][i2c.idf:096]: Found no i2c devices!
```

Is this the EEPROM? Is each I2C address a storage register?
I have never worked with a EEPROM and I am wondering why so many devices were found.

**Reply by syssi** [reply]
Interesting. Normally there is just a single i2c device at a single address. The short answer to 

> Is this the EEPROM? Is each I2C address a storage register?

is no! :-)

**Comment by B48D81EFCC** [comment]
Good news. I was able to connect a BME280 (temp + humidity + pressure) and a BH1750 (lux) successfully to the built-in esp32.
Next step is to make the wiring nicely.

**Reply by syssi** [reply]
Good job! Did you use the existing i2c bus or a second one?

**Reply by B48D81EFCC** [reply]
I used a separate bus.
The existing one looks suspicious with all these i2c devices on the bus

**Comment by B48D81EFCC** [comment]
I finished the upgrade.

I replaced the built-in board with a Seeed Studio XIAO ESP32S3.
I attached a SHT30 (Temp. and Humidity) and a BH1750 (Illuminance).

![image](https://github.com/user-attachments/assets/a1c148e0-b434-42c0-9b94-287a76e8816d)

**Reply by syssi** [reply]
Good job! Enjoy your unique ceiling light! :-)

**Comment by B48D81EFCC** [comment]
Hi,

I wanted to increase the brightness a bit and changed `max_power: 0.88` to `max_power: 1.00`.
Will this operate the LEDs outside their specs and decrease the lifetime significantly?

**Reply by syssi** [reply]
I don't know. I measured the voltage limits of the original firmware just to make sure to operate the device in the same range as the manfacturer.
